existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,309 @@
+* You can choose type for `b` only inside `func` definition, and you can't do it when you call `func` (we will discuss it further below).
+
+Here's a real-life example of using existential types in functions.
+Imagine a function that prints some logs while calculating the result.
+You may want to write logs into a console or into some file.
+In order to do it, you pass a logging function as an argument.
+
+```haskell
+func :: _ -> IO ()
+func log = do
+  username <- getUsername
+  log username
+
+  userCount <- getUserCount
+  log userCount
+
+getUsername :: IO String
+getUsername = undefined
+
+getUserCount :: IO Int
+getUserCount = undefined
+```
+
+What type should the log function have?
+Since it somehow logs both `String` and `Int`:
+
+```haskell
+log :: Show a => a -> IO ()
+```
+So, one might think, that adding a function with signature `a -> IO` and a constraint `Show a` would be enough to implement it.
+However, if we try:
+
+```haskell
+func :: Show a => (a -> IO ()) -> IO ()
+func log = ...
+
+main :: IO ()
+main = do
+  -- log to the console
+  func print
+
+  -- log to a logfile
+  func (appendFile "logfile.log" . show)
+```
+We will see some nasty compiler errors:
+```haskell
+  log username
+      ^^^^^^^^
+Couldn't match expected type `a` with actual type `String`...
+```
+
+We're getting this error because:
+1. `func`'s type variable `a` is universally quantified, therefore it can only be instantiated at the use-site.
+2. We're trying to instantiate it at `func`'s definition-site (`log username` is trying to instantiate the type variable `a` with `String` and `log userCount` is trying to instantiate the type variable `b` with `Int`).
+Because we want `a` to be instantiated at `func`'s definition-site (rather than its use-site), we need `a` to be existentially quantified instead.
+
+```haskell
+{-# LANGUAGE RankNTypes #-}
+
+func :: (forall a. Show a => a -> IO ()) -> IO ()
+```
+Now the compiler knows, that whatever function we pass to `func`, type variable `a` should be instantiated here, not outside.
+With this change we can instantiate `a` when we call log.
+
+```haskell
+func :: (forall a. Show a => a -> IO ()) -> IO ()
+func log = do
+  username <- getUsername
+  log @String username
+
+  userCount <- getUserCount
+  log @Int userCount
+```
+
+Note, that this type applications is unnecessary.
+Compiler is smart enough to choose appropriate types.
+
+#### Heterogeneous lists
+
+Now we understand that existential types allow you to introduce type variables without the necessity to specify them.
+Let's take a look how this is done in datatypes.
+You can introduce existentially quantified type variable in the datatype, by using `forall` on the left side of the constructor name. Existential types in datatypes allow you to have field in the datatype, without necessity to specify its type.
+
+```haskell
+data Elem = forall a. Elem a
```

Comment:
Is it just me or it would be much easier to introduce the concept of existential quantification by comparing a universally quantified and existentially quantified list?

Reply:
Not sure .... There is no such thing as existentially quantified list. We could only have a regular list, where each element contains existentially quantified variable. And in order to explain how it works, we need to explain existentially quantified datatype first

Reply:
Yeah, I mean a list where each element contains an existentially quantified variable. 

What I mean is that it is much easier to explain what existential quantification in Haskell is through an example. 

There are two examples: one with logging and the second one with list. 

So it would be good to start with either of them when exploring existential quantification. 

What I wanted to suggest is to focus on this list thing, since it feels simpler and then build up to the logging example (it is more useful) and the complete definition (it is a good way to structure what it _actually means).